### ISSUE TITLE
MSI buildbot patches version to avoid breaking distutils MSI builder, but does it in the wrong place

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5024]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5024|
|Type|defect|
|Created|2011-04-04 20:03:30Z|
|Branch|https://github.com/twisted/twisted/tree/msi-buildbot-5024-2|

As seen in [this build step](http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7-msi/builds/159/steps/shell_3/logs/stdio), we re-write the version by assigning it in `copyright.py`.

This is because of this unfortunate misfeature of distutils:
```
>>> from distutils.version import StrictVersion
>>> from twisted import __version__
>>> StrictVersion(__version__)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python2.6/distutils/version.py", line 40, in __init__
    self.parse(vstring)
  File "/usr/lib/python2.6/distutils/version.py", line 107, in parse
    raise ValueError, "invalid version number '%s'" % vstring
ValueError: invalid version number '11.0.0+r31541'
```
This raises a similar exception on version numbers like `11.0.0pre1`.

However, this assignment is done in `copyright.py`, which normally doesn't have a version declaration in it at all.  The actual version number is in `twisted._version`, and the string-ified short version that is exported as `twisted.__version__` is stringified in `twisted/__init__.py`.  The monkey-patching should therefore probably be done to `__init__.py`, so that it's already right when it's imported, and it matches in both places.

Unless the idea is that the versions _should_ differ, in which case it should be documented somehow.

Attachments:

* [5024.patch.version.for.bdist_msi.in.setup.py.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c2/9c245dd384542368fc86c61382de2df6bd58dbd5/e0c10937db6cb7df479ddb8b5044742e035c9eff.patch) (1132 bytes) - added by techtonik on 2011-04-04 20:50:08Z - 
* [5024.version.override.for.bdist_msi.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c2/9c245dd384542368fc86c61382de2df6bd58dbd5/361a0bd308f8b87f25d3f4614f20af6017233594.patch) (1378 bytes) - added by techtonik on 2011-04-05 15:44:14Z - version 2
* [5024_3.version.monkeypatch.for.bdist_msi.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c2/9c245dd384542368fc86c61382de2df6bd58dbd5/4f7df130b262e81da3c4d9417eb37c696c78905e.patch) (1904 bytes) - added by techtonik on 2011-04-05 21:30:29Z - version 3
* [5024.4.bdist_msi.version.normalize.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c2/9c245dd384542368fc86c61382de2df6bd58dbd5/e218ec7edff72d7d8c26af348be132c6e4b33efd.patch) (2538 bytes) - added by techtonik on 2011-04-15 08:13:27Z - version 4: moved fix to twisted\python\dist.py
* [5024.6.bdist_msi.version.normalize.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9c2/9c245dd384542368fc86c61382de2df6bd58dbd5/656285a19cb035d8e373e136d8bf821868e07206.patch) (3604 bytes) - added by techtonik on 2011-04-15 09:53:03Z - improved comments, add simple test

<details><summary>Searchable metadata</summary>

```
trac-id__5024 5024
type__defect defect
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__branches_msi_buildbot_5024_2 branches/msi-buildbot-5024-2
branch_author__thijs__techtonik thijs, techtonik
status__new new
resolution__None None
component__core core
keywords__ 
time__1301947410000000 1301947410000000
changetime__1317670727000000 1317670727000000
version__None None
owner__techtonik techtonik
cc__oubiwann cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|techtonik commented|
|-|-|

I am not sure if it works (still no visual studio on this machine), but setup.py looks like the most appropriate place for patching.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to techtonik|
|-|-|

Replying to [techtonik](#note_2):
> I am not sure if it works (still no visual studio on this machine), but setup.py looks like the most appropriate place for patching.

Thanks for the patch.

  1. This is quite possibly a better place to do it, but it needs some automated test coverage.  Possibly somewhere in `twisted.test.test_release`?  You could add this coverage by writing a function and testing it against `StrictVersion` at first, we don't necessarily need end-to-end coverage of `setup.py`, that's what the buildbots are for.
  2. Also, the message that is emitted should link to some explanation of why this needs doing and what the version is being changed from/to.
  3. If we fix it this way, there needs to be a separate ticket filed for removing the junk step from the buildbot config.  And this will need to be noted on the ReleaseProcess page, in case somebody needs to do an MSI maintenance release of an older version, they will need to backport the fix.

As long as we're cleaning this area up, I suspect that we could add some extra information to the MSI version number as well; perhaps x.y.z.subversionrev?  It would need some checking with `StrictVersion` first, of course.

Again, thanks for the patch; glad to see you submitting stuff to Twisted, anatoly :).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

I'll put your patch in a branch so it can be tested against the buildslaves.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

(In [31549]) Branching to 'msi-buildbot-5024'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [31550]) apply 5024.version.override.for.bdist_msi.patch, refs [#5024](https://github.com/twisted/twisted/issues/5024)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra set owner to techtonik<br>@thijstriemstra set status to `new`|
|-|-|

Applied the patch and added an @since marker. Assigning back to techtonic (since the ticket wasn't put up for review yet).

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|techtonik commented|
|-|-|

Thanks for putting this to the test, thijs. Where can I find buildslaves that work on branches? I could find trunk/ bots only, and [Combinator page](http://divmod.org/trac/wiki/DivmodCombinator) describing branching the process is offline for some reason.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|techtonik commented|
|-|-|

The main thing I forgot is that this change can not be tested without modification to buildslave configuration.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|techtonik commented|
|-|-|

Replying to [glyph](#note_3):
> 
>   1. This is quite possibly a better place to do it, but it needs some automated test coverage.  Possibly somewhere in `twisted.test.test_release`?  You could add this coverage by writing a function and testing it against `StrictVersion` at first, we don't necessarily need end-to-end coverage of `setup.py`, that's what the buildbots are for.

I've added a proper distutils override in 2nd version of the patch just attached. Any ideas how to test this? I'm still blocked by [#3586](https://github.com/twisted/twisted/issues/3586) to test even manually, though freshly found `--skip-build` option for `bdist_msi` should help.

>   2. Also, the message that is emitted should link to some explanation of why this needs doing and what the version is being changed from/to.

version 2 doesn't emit any messages. It should limit hack scope to `bdist_msi`, but now I can see that `bdist_msi` calls `build` and `install` inside. Though it is still better than patching sources right before packing distributive. I'll try to come up with a better third version.

>   3. If we fix it this way, there needs to be a separate ticket filed for removing the junk step from the buildbot config.  And this will need to be noted on the ReleaseProcess page, in case somebody needs to do an MSI maintenance release of an older version, they will need to backport the fix.

ReleaseProcess can link to the comment in this separate ticket describing the process.

> As long as we're cleaning this area up, I suspect that we could add some extra information to the MSI version number as well; perhaps x.y.z.subversionrev?  It would need some checking with `StrictVersion` first, of course.

Not supported - it is copied to [MSI ProductVersion](http://msdn.microsoft.com/en-us/library/aa370859(v=vs.85).aspx) as is.

> Again, thanks for the patch; glad to see you submitting stuff to Twisted, anatoly :).

Just because I've got some free time to learn Twisted if I want to get speaking IRC monkeys in Blender. It is not at all that I care about [highscores](http://twistedmatrix.com/highscores/). =)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

Replying to [techtonik](#note_8):
> Thanks for putting this to the test, thijs. Where can I find buildslaves that work on branches? I could find trunk/ bots only, and [Combinator page](http://divmod.org/trac/wiki/DivmodCombinator) describing branching the process is offline for some reason.

You can test this against any of the win32 builders, like [windows7-64-py2.7-msi](http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7-msi). Submit the 'Force Build' form and point the 'branch' field to `branches/msi-buildbot-5024` (+ the name/reason fields, other fields can be empty).

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Replying to [techtonik](#note_9):
> The main thing I forgot is that this change can not be tested without modification to buildslave configuration.

You can ask exarkun on IRC, he manages the buildfarm afaik.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|techtonik commented|
|-|-|

Wow. That's really convenient. However, the build [failed with bzr exception](http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7-msi/builds/161).

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|techtonik commented|
|-|-|

All right - this one built o.k., but it still includes copyright.py patch.

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_15"></a>|techtonik commented|
|-|-|

I mean [this one](http://buildbot.twistedmatrix.com/builders/winxp32-py2.6-msi/builds/358).

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|techtonik removed owner|
|-|-|

version 3 is the properly monkey patch to limit the scope of version modification to calls from `bdist_py` only. Slighly hardcore. Choose which patch is better. I still have no idea how to test this, but I was able to manually test that fix works with:
```
python setup.py bdist_msi --skip-build
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to techtonik|
|-|-|

Thanks for the patches, techtonik.

  1. A start on testing might be to move the customized `bdist_msi` command class into `twisted/python/dist.py` which is where we keep much of our other distutils hackery.  Notice that we're not calling `distutils.core.setup` in our setup.py, we're calling  calling `twisted.python.dist.setup`.  So you may even be able to implement this change without modifying setup.py at all.  Which hopefully means unit tests can cover all of the changes.
  1. Can we have a unit test that just calls `bdist_msi_patch_version.run` and asserts something about the consequences?  Or perhaps something that calls `twisted.python.dist.setup` with some arguments to a fake mini project and assert something about the outcome of that?

**Comment by twisted-trac** [comment]
|<img alt="techtonik's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_18"></a>|techtonik commented|
|-|-|

Replying to [exarkun](#note_17):
>   1. A start on testing might be to move the customized `bdist_msi` command class into `twisted/python/dist.py`

Done.

>   1. Can we have a unit test that just calls `bdist_msi_patch_version.run` and asserts something about the consequences?  Or perhaps something that calls `twisted.python.dist.setup` with some arguments to a fake mini project and assert something about the outcome of that?

I've made a test that ensures the target function is patched. Testing the behavior of patched function is harder - the original distutils.bdist_msi.run() method should be run, and it builds the distribution writing many files. It might be possible to do a fake mini project and execute build process with --dry-run flag, but this will take some time to do.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set owner to techtonik|
|-|-|

1.The Python 2.3 check touched in the patch can go, or should  be replaced with a >= 2.4 check, python2.3 isn't supported anymore in twisted. 

```
if sys.version_info[:3] < (2, 3, 0):
```

 2. The class docstrings in t.p.dist should be on 3 lines
 3. I don't like the print saying it 'may' fail. If it's something we can fix later, then a twisted ticket reference should be included (in a comment above it perhaps)
```
print "Twisted Warning: cmdclass override from setup.py may fail"
```

btw, could you make new patches against the branch? thanks.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_23"></a>|@thijstriemstra commented|
|-|-|

(In [32531]) Branching to 'msi-buildbot-5024-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra commented|
|-|-|

(In [32532]) apply 5024.6.bdist_msi.version.normalize.patch, refs [#5024](https://github.com/twisted/twisted/issues/5024)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Up for review, [build results](http://buildbot.twistedmatrix.com/builders/windows7-64-py2.7-msi/builds/305).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

Thanks for picking this up thijs.

distutils is pretty nasty, and Windows is worse, so I shouldn't be too surprised that this isn't a very enjoyable patch.

Here are some bad things about it:

  1. The stack inspection in `monkey_get_version` makes it almost impossible to unit test in a useful way.  A unit test simply isn't going to have bdist_msi.py near the top of its stack.
  1. It's not clear why there's a conditional in `monkey_get_version` at all.  Why is the full version ever appropriate?  Testing with the conditional removed the build seems just as successful.
  1. The svn information isn't always separated from the main version string by a "-".  A "+" is sometimes used instead (best I can tell, this hinges on whether or not setuptools is installed).
  1. The import of `distutils.command.bdist_msi` in `twisted.python.dist` makes that module unimportable except on Windows.
  1. The changes to how the `'cmdclass'` entry of `kw` in `get_setup_args` is initialized seems extraneous.  The new print seems to be as well.  Perhaps there is a problem, but it is not related to building Windows MSI packages - so separate ticket.
  1. The unit tests don't assert much of interest.  The hash of the `get_version` method is of tertiary importance at best.  I don't think there's any guarantee the hash will be the same even if the monkey patch _doesn't_ take place.  And it doesn't really tell us that bdist_msi will succeed.
  1. The unit test also needs to skip except on Windows.

That said, the patch _does_ appear to fix the problem - although the MSI build linked above doesn't demonstrate it, since it is based on a buildbot configuration which includes the workaround for this issue - and that's a big upside.

If the platform issues are fixed, and the unit test is changed to do something directly with `StrictVersion` (since that seems to be the unit which is having difficulties), and the unneeded conditional is removed (or explained) then I think the patch will be acceptable.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to techtonik|
|-|-|
